numba/targets/listobj.py

```diff
@@ -1031,57 +1032,85 @@ def list_reverse_impl(lst):
 # -----------------------------------------------------------------------------
 # Sorting
 
-_sorting_init = False
-
-def load_sorts():
-    """
-    Load quicksort lazily, to avoid circular imports across the jit() global.
-    """
-    g = globals()
-    if g['_sorting_init']:
-        return
-
-    def gt(a, b):
-        return a > b
-
-    default_sort = quicksort.make_jit_quicksort()
-    reversed_sort = quicksort.make_jit_quicksort(lt=gt)
-    g['run_default_sort'] = default_sort.run_quicksort
-    g['run_reversed_sort'] = reversed_sort.run_quicksort
-    g['_sorting_init'] = True
-
-
-@lower_builtin("list.sort", types.List)
-@lower_builtin("list.sort", types.List, types.Boolean)
-def list_sort(context, builder, sig, args):
-    load_sorts()
-
-    if len(args) == 1:
-        sig = typing.signature(sig.return_type, *sig.args + (types.boolean,))
-        args = tuple(args) + (cgutils.false_bit,)
-
-    def list_sort_impl(lst, reverse):
-        if reverse:
-            run_reversed_sort(lst)
+def gt(a, b):
+    return a > b
+
+sort_forwards = quicksort.make_jit_quicksort().run_quicksort
+sort_backwards = quicksort.make_jit_quicksort(lt=gt).run_quicksort
+
+arg_sort_forwards = quicksort.make_jit_quicksort(is_argsort=True,
+                                                 is_list=True).run_quicksort
+arg_sort_backwards = quicksort.make_jit_quicksort(is_argsort=True, lt=gt,
+                                                  is_list=True).run_quicksort
+
+
+def _sort_check_reverse(reverse):
+    if isinstance(reverse, types.Omitted):
+        rty = reverse.value
+    elif isinstance(reverse, types.Optional):
+        rty = reverse.type
+    else:
+        rty = reverse
+    if not isinstance(rty, (types.Boolean, types.Integer, int, bool)):
+        msg = "an integer is required for 'reverse' (got type %s)" % reverse
+        raise errors.TypingError(msg)
+    return rty
+
+
+def _sort_check_key(key):
+    if isinstance(key, types.Optional):
+        msg = ("Key must concretely be None or a Numba JIT compiled function, "
+               "an Optional (union of None and a value) was found")
+        raise errors.TypingError(msg)
+    if not (isinstance(key, (types.NoneType, types.Omitted)) or key is None or
+            isinstance(key, types.Dispatcher)):
+        msg = "Key must be None or a Numba JIT compiled function"
+        raise errors.TypingError(msg)
+
+
+@overload_method(types.List, "sort")
+def ol_list_sort(lst, key=None, reverse=False):
+
+    _sort_check_key(key)
+    _sort_check_reverse(reverse)
+
+    if (isinstance(key, (types.NoneType, types.Omitted)) or key is None):
```

Comment:
use `cgutils.is_nonelike()` instead

Reply:
Why is this in `cgutils`? It's typing not codegen?

Reply:
good point. there's no typing util either. will need to sort that out later.